### ISSUE TITLE
Declaratively Configurable Markup

### ISSUE THREAD

**Issue by trieloff** [adobe/helix-pipeline#516]
This is mainly related to the `helix-pages` use case, but should be implemented straight in the pipeline.

The HTML generated by the pipeline is intentionally minimalistic. As @anfibiacreativa pointed out during the hackathon, minimalistic markup can lead to maximalistic CSS selector expressions and it would be great to give developers a way to influence the generated markup.

With the `VDOMTransformer.match` API, this ability exists, but it comes with three drawbacks:
- lacking a `pre.js` in `helix-pages`, it cannot be used in the most common scenario
- being an imperative API, it cannot be safely run in a shared environment
- the API requires advanced understanding of the `unifiedjs`, `hyperscript` and DOM APIs, and requires recursion, which can pose a hurdle for most developers

In practice, most scenarios require one of the following:
1. **wrap** the generated HTML element in one or more developer-defined elements
2. **inject** `class` attribute values into the `class`attribute of the generated HTML element
3. **add** other attributes and values into the generated HTML element

These use cases can be solved using a similar approach we use for [query definitions](https://github.com/anfibiacreativa/helix-norddal/blob/master/helix-index.yaml).

Imagine a `markup.yaml` (https://github.com/adobe/helix-shared/issues/190) file like this:

```yaml
#
# markup configuration for theblog
#
version: 1
​
markup:
  images-in-gallery:
    match: section.is-gallery image
    wrap: <div class="gallery-image">
  external-links:
    match: link[href!="https://"]
    class: external
  last-section:
    match: section:last-of-type
    attribute:
      data-type: footer
```

The `match` expression would work on the MDAST tree, and the `wrap`, `class`, or `attribute` directives would apply to the generated DOM nodes.

**Comment by davidnuescheler** [comment]
i think it would be great to support css selectors level 4

**Comment by trieloff** [comment]
> i think it would be great to support css selectors level 4

But then they'd work on the DOM, not the MDAST, correct?

**Comment by tripodsan** [comment]
> The `match` expression would work on the MDAST tree, and the `wrap`, `class`, or `attribute` directives would apply to the generated DOM nodes.

how would that work? this isn't very intuitive. Wouldn't it be better to match the DOM instead?

**Comment by davidnuescheler** [comment]
i definitely agree with @tripodsan that they should be added to DOM, specifically as i think that this is a feature that i think should be mostly useful for helix-pages as the only way to adapt the markup and i don't think that the MDAST would be something that i would want to be understood by a helix-pages developer

**Comment by trieloff** [comment]
That's the default since eb3d314

**Comment by ramboz** [comment]
I could imagine 2 hooks in the pipeline that respectively:
- match on the MDAST and modify the tree accordingly before HTML processing
- match on the HTML and manipulate the resulting DOM

depending on what `type` is used on the markup config.

> `wrap: <div class="gallery-image">`

I wonder if this couldn't become an XSS attack vector. Also using HTML in the markup feels like we are adding another level of complexity/expertise to be able to use the API (CSS+YAML+HTML).

I would propose to use an emmet/zen-coding like notation instead that basically re-uses CSS selectors,:
```yaml
wrap: .gallery-image # => <div class="gallery-image">…</div>
wrap: section.gallery-image # => <section class="gallery-image">…</section>
wrap: foo[bar="baz"] .qux # => <foo bar="baz"><div class="qux">…</div></foo>
```

**Comment by trieloff** [comment]
The emment notation may make things easier. I don't see the XSS vector, given that the only people capable of adding HTML are authors.

**Comment by ramboz** [comment]
@trieloff They may be adding unsafe HTML, which, if sanitization is not enforced, basically exposes the XSS in the resulting DOM.

On a side note, I started wondering if all of this shouldn't be part of before/after hooks in helix pages, instead of directly in helix. The reasoning being that anyone directly using helix would know how to leverage the .pre.js or HTL to achieve the same. What do you think?

**Comment by tripodsan** [comment]
@ramboz I like the declarative wrap a lot...

also see https://github.com/adobe/helix-pipeline/pull/529#pullrequestreview-316642892

how would we describe a `replace` ?

**Comment by trieloff** [comment]
Authors are not attackers. If an author wants to use JavaScript in a Helix (Pages) site, there are already plenty of ways:
- use an HTML block in Markdown
- replace the Markdown file with a similarly named HTML file
- add it to footer.html or header.html

Now we are adding another way: change markup.yaml, but this is not a substantial change from the status quo.

@Davidnuescheler what about adding a statement like "authors are not XSS attackers to the manifest?

**Comment by ramboz** [comment]
> how would we describe a replace ?

@tripodsan I'd probably just use `replace` instead of `wrap`, I guess. I think the emmet notation would be the same, except you don't need the `{$#}` in the last example as we don't need to have a placeholder for existing nodes

> Authors are not attackers.

@trieloff I don't consider them to be. But I also don't expect them to be tech-savvy if they are using the `markup.yaml` file. So it's easy for them to inject JS with the HTML that would leave doors open for a later attacker on the page.

This isn't an issue though if we move from HTML to emmet-style notation in the markup.yaml, as there is no possibility to inject scripts there.

**Comment by trieloff** [comment]
Btw. I don’t think a replace operation is really practical or needed. If you want to replace content with something static, you can do this with CSS. Let’s not add replace for now.

**Comment by trieloff** [comment]
Using markup.yaml has little to do with tech savvyness and more with using the available tools. In Helix Pages, you don’t have a Pre JS, so you don’t use it, technical or not.

**Comment by tripodsan** [comment]
> Btw. I don’t think a replace operation is really practical or needed.

how would you then transform `<svg class="icon adobe" src="/icon/adobe.svg"></svg>` to `<img class="icon adobe" src="/design/icons/adobe.png">` ?

>  If you want to replace content with something static, you can do this with CSS.

how?

**Comment by trieloff** [comment]
```css
.mything {
display: none;
}

.mything:after {
display: block;
content: "My Thing";
}
```

This doesn’t allow for dynamic markup generation, but that wasn’t part of the original scope anyway.

**Comment by kptdobe** [comment]
I like the proposed API but not the fact that this ends up into a YAML file. Main reasons: 
- you will always end up in a case where the `wrap` / `inject` / `add` is not enough
- how do you deal with the `if` ? The incoming DOM is controlled by the author, he can write something unexpected, you need life-guards, you cannot handle those

I would rather do something like a `markup.prepare.js` (loaded from the content too for helix-pages) which would contain: 

```
markup.match('section.is-gallery image').wrap('<div class="gallery-image">');
markup.match('link[href!="https://"]').class('external');
markup.match('section:last-of-type').('data-type: footer')
const authorURL = ... // compute an author url to fetch the html from
If (authorURL) {
   markup.replace('section.authror', get(authorURL));
}
```

**Comment by ramboz** [comment]
@acapr I think the solution addresses a different audience. My understanding so far was that we would have a solution for non-JS coders and that would let them do basic markup customizations to perform more advanced CSS styling that in the basic helix pages examples. I'm looking at a small design agency that has design, CSS and HTML expertise. JS capabilities would be limited to integrating JS libs required for slideshows, etc. but no actual JS coding.

If you have JS expertise and are able to use a `markup.prepare.js`, then why not just do most of it in the `.pre.js` in the first place?

**Comment by kptdobe** [comment]
The `pre.js` is not available in `helix-pages` and I think it should not. We should just open a another extension point.

The `markup.yaml` will be grow and grow to support more features and we'll quickly become a huge machinery. theblog we are trying to build is super basic / simple in terms of design and requires some really basic capabilities the `markup.yaml` will not offer.

**Comment by kptdobe** [comment]
Your small agency will be stuck very quickly.
To build something useful here, we should rather mimic some sites, do the actual work (like we do for theblog), I am almost sure we'll find this too limited.

**Comment by ramboz** [comment]
I guess I'm missing some background then… where are these constraints documented? Is there a background discussion to track? What are the exact use cases we want to achieve?

The `markup.yaml`'s relation to the blog is also new to me here. I don't really see this documented in the issue. Maybe the initial idea does match the current expectations anymore.

> Your small agency will be stuck very quickly.

Yes. that is the expectation. But the goal, in my mind, was to address a small 1-2 people (with only design/CSS/HTML expertise) that will sell a "theme" for say $20 on themeforst or similar. I'm looking at someone who will spend max 1 week creating a quick design and then the matching CSS to be applied on top of an out-of-the-box helix pages project. The goal is just to adjust layout, colors, maybe a bit of SEO and A11y. Mostly "white-labelling". So we let you wrap HTML as you need, add attributes and inject CSS classes. If you need more, then I think you definitely outgrow the `markup.yaml` and you are then in the "medium agency" category. For one, the `markup.yaml` has no `VDOMTransformer.match` equivalent to replace content. We just wrap.

That medium agency definitely also needs an extension mechanism, and I agree with you that the `markup.yaml` will not be enough. I would also expect a more JS based extension mechanism for them. But I don't feel this was the targeted audience when I look at the initial issue description.
I could definitely imagine a `markup.transform.js` API that exposes `VDOMTransformer.match` for instance.

If I look at a simple wordpress blog as a reference, and the theming options available for those, you'll quickly find:
- basic CSS theming done on top of Bootstrap (or equivalent), usually in a $10~$30 range
- more advanced theming including various customizable layouts, JS libraries for parallax/slideshows/etc. and that output custom HTML, usually in the $50~$100 range. This will include PHP, CSS, JS, etc.
- serious extensions that enhance the default wordpress blog and add new features like product catalogs, ecommerce, etc.

`markup.yaml` would address the basic use cases, and `markup.transform.js` could probably address the more advanced ones. But I feel the 2 may be better addressed in separate tickets with clear separate use cases, tests and documentation.

**Comment by kptdobe** [comment]
I was not really of this ticket before ;)
I just trying to implement "theblog" (see https://github.com/davidnuescheler/theblog/issues) and I am stuck in multiple places. This ticket looks the closest to what I am trying to do. 
The pb is that this ticket does not cover something specific, just a vague guess that we will be able to cover more. But when trying to face reality (like rendering this markdown https://github.com/davidnuescheler/theblog/blob/alexonedrive/content/posts/creating-adobe-experience-platform-pipeline-with-kafka.md into this https://theblog.adobe.com/creating-adobe-experience-platform-pipeline-with-kafka/) which is supposed to be straight forward, you see the edge cases and you see them pretty quickly.
While I see some benefits, I think it is not enough, even for a basic use cases.

Something you can hardly express here is that the markdown will be written by an author that will make mistakes (an extra line break could be one). We need to make sure the DOM "enhancement" logic is strong enough.

But maybe you are right, some of this could be tackled in another issue.

**Comment by ramboz** [comment]
@kptdobe Happy to discuss chat on the edge cases and brainstorm around a solution, and we can then maybe define a proper issue for it if needed.

As far as "straight forward" goes, if you want to have a properly accessible blog with good SEO, it will quickly become not-so-straight-forward in my (non-helix) experience and a lot of hardened logic will need to go into the HTL templates I think. I would only expect `markup.yaml` to be used to theme (white label) the blog for your company once you have the logic in place, but I would not rely on it to generate the right markup though.

The hardening you mention around content, I think, needs to go in the HTL and pre.js, rather than being done through a theming extension mechanism.

**Comment by kptdobe** [comment]
Maybe I should re-explain what I am doing and why I think this is not enough.

I started with a given Markdown, https://github.com/davidnuescheler/theblog/blob/alexonedrive/content/posts/creating-adobe-experience-platform-pipeline-with-kafka.md , already designed by @davidnuescheler here https://github.com/davidnuescheler/theblog/blob/alexonedrive/style.css. Goal here is to use only helix-pages.

My first "worry" is that the .md file contains several thematic breaks, mainly around the first image in order to be style it (fullscreen image).
This is something that would be fixable with the current proposal, wrap the first elements into more divs.
But next "worry" is: what if the .md file does not exactly follow the expected pattern, i.e. title first, hero image second (we have the case of existing blog entries with images). The majority of the cases we could imagine could be simply be handled by some "ifs", if image is not there, if missing the first title and allow to render something nice if the author makes mistake. This might be solvable with CSS (thus with current proposal) but it gets already more fuzzy.

Assume this is would be doable with `markup.yaml`.

Next next "worry": the author block needs a little bit of logic to be treated - extract the author name, compute its avatar image url and inject it in the DOM, which I have done for the moment clientside (see https://github.com/davidnuescheler/theblog/issues/8) which is I think wrong because of caching and also we do not have the support in helix-pages for this use case. Nothing complicated but based on your description, it forces me to go from "small agency" to "medium agency" category. Fair enough, I now need to implement a `pre.js` (assuming it is supported by helix-pages). But I have already stuff in the `markup.yaml` ! Now I start having to maintain 2 different concepts to support my DOM enhancements.

**Comment by ramboz** [comment]
It seems to me that all the issues you are describing are semantic HTML structure issues, not styling issues. The `markup.yaml` (maybe the name is misleading?), in my opinion, can only address styling issues.
You can't clean up HTML and enforce proper accessible and SEO-friendly markup by just adding a class, wrapping an element, and maybe adding 1 or 2 attributes. You really need to manipulate the HTML structure, change tags, etc. If your HTL does not already output the proper semantic structure, then the `markup.yaml` won't be able to help much

What you describe, I think, should all be rather covered in the HTL for that blog page and you'd have a different HTL for "blog" entries than regular pages. Basically, I'd expect the former to follow https://schema.org/BlogPosting while the latter follows https://schema.org/WebPage (or one of its sub-types).

Once you have the proper SEO/a11y-friendly markup, the user can use the yaml to white-label if out-of-the-box structure is not flexible enough. You, as a developer, should not have to touch the `markup.yaml` unless you are theming the blog for a new customer, I think. You would do all in the business logic in the HTL and `pre.js`

**Comment by trieloff** [comment]
@kptdobe and @ramboz thanks for the discussion and perspective. I'll add my point of view in general:

1. we should restrict ourselves to something that is entirely declarative, so that it works in Helix Pages (where we can't run untrusted code, and we should treat any imperative code as not trustworthy)
2. we can expand the scope of the available actions or transformations over time, but it makes sense to start small
3. I don't see how adding an `if` construct would allow us to do things that are categorically different from what we can do with CSS selectors (which are just a DSL for if-statements on a DOM)

**Comment by trieloff** [comment]
As for the metadata parsing challenge that you are experiencing with theblog, @kptdobe, this might be the wrong hammer to drive in this particular screw, so we might want to discuss this in a separate issue.

**Comment by tripodsan** [comment]
from helix-chat:

> talking about HelpX and some of their requirements. One interesting use case is an idea of ‘renditions’ for different audiences. You might have a page like https://helpx.adobe.com/fonts/using/open-type-syntax.html but there’s an ability to have a totally different display using a selector like: https://helpx.adobe.com/fonts/using/open-type-syntax.ug.html
Can Helix support something similar today? I was trying to think of what’s available in pipelines and HTL, but couldn’t get a clear idea of how to do this.

> well, the helix pipeline is especially built with selectors in mind. so you can have a html.htl and ug_html.htl to render the resource differently.
> with helix-pages this is currently not possible, as heliix-pages controlls the rendering.
> however, maybe with the new declarative markup feature (https://github.com/adobe/helix-pipeline/issues/516) we could take the selector into account in the config (or load a different config completely), so you can tweak the layout for different selectors.

**Comment by rofe** [comment]
@ramboz should we document the final state of this feature somewhere outside of this issue and PR? The latter may be a bit hard to follow for the uninitiated.

**Comment by ramboz** [comment]
@rofe Sure, happy to. What is the best practice for this?

**Comment by rofe** [comment]
> What is the best practice for this?

There isn't any 😄 

Helix documentation is scattered across our many repos at the moment, and it's gotten worse since we decided to discontinue https://www.project-helix.io/... 

For `helix-pipeline`, there is a [docs folder](../docs) for instance, but it seems to talk mostly about schemas. The best place is probably [README.md](../README.md) for now.

Personally, I am in favor of starting to collect Helix documentation in a central place:
https://github.com/adobe/helix-home/tree/master/docs

**Comment by trieloff** [comment]
I think it would be ideal to have some examples here: https://github.com/adobe/helix-shared/blob/master/src/schemas/markupmapping.description.md so that it can be used in the reference documentation.

Ideally consisting of a snippet of Markup, a YAML snippet and the resulting HTML.

**Comment by ramboz** [comment]
PR: https://github.com/adobe/helix-shared/pull/267

**Comment by rofe** [comment]
Yet another repo ...
